File: 140_Coin Change.py
"""
140. Coin Change
Use DP to find the fewest coins to make amount (unbounded knapsack).
"""
from typing import List

def coin_change(coins: List[int], amount: int) -> int:
    INF = 10**9
    dp = [0] + [INF] * amount
    for c in coins:
        for a in range(c, amount+1):
            dp[a] = min(dp[a], dp[a-c] + 1)
    return -1 if dp[amount] == INF else dp[amount]

if __name__ == '__main__':
    print('coinChange([1,2,5],11) =', coin_change([1,2,5],11))
    print('coinChange([2],3) =', coin_change([2],3))

# Time Complexity: O(amount * n_coins)
# Space Complexity: O(amount)

File: 140_Coin Change.cpp
/*
140. Coin Change
Given coin denominations and amount, find the fewest coins needed to make up that amount (unbounded knapsack / DP).
Return -1 if not possible.
*/
#include <bits/stdc++.h>
using namespace std;

int coinChange(vector<int>& coins, int amount) {
    const int INF = 1e9;
    vector<int> dp(amount+1, INF);
    dp[0] = 0;
    for (int c : coins) {
        for (int a = c; a <= amount; ++a) dp[a] = min(dp[a], dp[a-c] + 1);
    }
    return dp[amount] == INF ? -1 : dp[amount];
}

int main() {
    vector<int> coins{1,2,5};
    cout << "coinChange([1,2,5],11) = " << coinChange(coins, 11) << "\n"; // 3
    vector<int> coins2{2};
    cout << "coinChange([2],3) = " << coinChange(coins2, 3) << "\n"; // -1
    return 0;
}

// Time Complexity: O(amount * n_coins)
// Space Complexity: O(amount)
